### PR TITLE
Use winsock functions for Win32 platform.

### PR DESCRIPTION
**Pull request by Biswa96**

### CODE REVIEW

File: cpan/Socket/Socket.xs [hunk 1]
```diff
@@ -51,105 +51,6 @@
 # include <ws2tcpip.h>
 #endif
 
-#ifdef WIN32
```
Comment (Grinnz) [inline cpan/Socket/Socket.xs]: This file is maintained on CPAN, and cannot be edited here. Its bugtracker is here: https://rt.cpan.org/Public/Dist/Display.html?Name=Socket

Comment (Biswa96) [reply cpan/Socket/Socket.xs]: How can I send patch file there without creating another account? Also there are bugs from 9 years ago, does author maintain it properly?

Comment (Grinnz) [reply cpan/Socket/Socket.xs]: Yes, the last release was three days ago. You can [report bugs to RT via email](https://rt.cpan.org/NoAuth/RT/CPAN/Usage.html), or login to it using your github account (you need to click "Log out guest user" to access the login).

Comment (Biswa96) [reply cpan/Socket/Socket.xs]: Thanks, I shall try.

Comment (Biswa96) [reply cpan/Socket/Socket.xs]: Is there any way to clone that Socket module or I have to make the patch file from tarball?

**Comment by sisyphus** [comment]
With mingw-w64 port of gcc-11.1.0, runtime version 9, the fix-winsock branch is failing for me in cpan/Socket/Socket.xs:
```
Socket.xs:101:12: error: static declaration of 'inet_pton' follows non-static declaration
  101 | static int inet_pton(int af, const char *src, void *dst)
      |            ^~~~~~~~~
In file included from ..\..\lib\CORE/sys/socket.h:18,
                 from ..\..\lib\CORE/win32.h:487,
                 from ..\..\lib\CORE/win32thread.h:4,
                 from ..\..\lib\CORE/perl.h:3342,
                 from Socket.xs:3:
c:\_64\winlibs-gcc-1110\mingw64\x86_64-w64-mingw32\include\ws2tcpip.h:408:32: note: previous declaration of 'inet_pton' with type 'INT(INT,  const CHAR *, void *)' {aka 'int(int,  const char *, void *)'}
  408 | WINSOCK_API_LINKAGE INT WSAAPI InetPtonA(INT Family, LPCSTR pStringBuf, PVOID pAddr);
      |                                ^~~~~~~~~
Socket.xs:123:20: error: conflicting types for 'inet_ntop'; have 'const char *(int,  const void *, char *, socklen_t)' {aka 'const char *(int,  const void *, char *, int)'}
  123 | static const char *inet_ntop(int af, const void *src, char *dst, socklen_t size)
      |                    ^~~~~~~~~
In file included from ..\..\lib\CORE/sys/socket.h:18,
                 from ..\..\lib\CORE/win32.h:487,
                 from ..\..\lib\CORE/win32thread.h:4,
                 from ..\..\lib\CORE/perl.h:3342,
                 from Socket.xs:3:
c:\_64\winlibs-gcc-1110\mingw64\x86_64-w64-mingw32\include\ws2tcpip.h:401:35: note: previous declaration of 'inet_ntop' with type 'const CHAR *(INT,  const void *, CHAR *, size_t)' {aka 'const char *(int,  const void *, char *, long long unsigned int)'}
  401 | WINSOCK_API_LINKAGE LPCSTR WSAAPI InetNtopA(INT Family, LPCVOID pAddr, LPSTR pStringBuf, size_t StringBufSize);
      |                                   ^~~~~~~~~
gmake[1]: *** [Makefile:337: Socket.o] Error 1
```

With the same compiler and runtime version 8, perl builds ok, but a couple of tests fail:

```
../cpan/IO-Socket-IP/t/04local-client-v6.t                         (Wstat: 6400
Tests: 0 Failed: 0)
  Non-zero exit status: 25
  Parse errors: No plan found in TAP output
../cpan/IO-Socket-IP/t/05local-server-v6.t                         (Wstat: 5632
Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 22
  Parse errors: No plan found in TAP output
```

With Visual Studio 2019, I get the following test failures:

```
../cpan/IO-Socket-IP/t/04local-client-v6.t                         (Wstat: 2304
Tests: 0 Failed: 0)
  Non-zero exit status: 9
  Parse errors: No plan found in TAP output
../cpan/IO-Socket-IP/t/05local-server-v6.t                         (Wstat: 2304
Tests: 0 Failed: 0)
  Non-zero exit status: 9
  Parse errors: No plan found in TAP output
../dist/IO/t/io_sock.t                                             (Wstat: 32768
 Tests: 25 Failed: 1)
  Failed test:  25
  Non-zero exit status: 128
  Parse errors: Bad plan.  You planned 26 tests but ran 25.
```
Also ../dist/Storable/t/code.t hanged during the running of the test suite.

I recognize the Socket.xs compilation failure from my own endeavours in #18911 - a redefining of inet_pton() and inet_ntop() after they've already been defined in ws2tcpip.h
But I've not seen the other failures anywhere before.

@Biswa96 I think I've correctly got hold of the fix-winsock branch. I've looked at cpan/IO/poll.h, and it's exactly as per your patch provided above.
I do like that you've tried to fix up the specs in the win32/config* files. I hope this can be made to work.
But, as it stands for me at the moment, all that's working for me are the latest patches I posted in #18911.
I think I could replace my poll.h patch with the one you've provided here - though I haven't finished verifying that yet.

**Comment by jkeenan** [comment]
There has been no further discussion of this p.r. since August 2021 and it has acquired merge conflicts.

I recommend that it be closed within 30 days if the merge conflicts are not corrected or if there is no further discussion.

**Pull request closed without merging**